[bug] Conan created toolchain accesses an undefined variable

### Environment details

* Operating System+version: mac and ubuntu in GitHub CI
* Compiler+version: -
* Conan version: 2.1.0
* Python version: 3.8

https://github.com/friendlyanon/cmake-init/actions/runs/8003305023/job/21858340788?check_suite_focus=true

### Steps to reproduce

Configure a project with `uninitialized` warning turned on and turned into an error.

### Logs

```
CMake Error (dev) at conan/conan_toolchain.cmake:44 (foreach):
  uninitialized variable 'CMAKE_CONFIGURATION_TYPES'
Call Stack (most recent call first):
  /usr/local/share/cmake-3.28/Modules/CMakeDetermineSystem.cmake:170 (include)
  CMakeLists.txt:5 (project)
This error is for project developers. Use -Wno-error=dev to suppress it.
```

--- a/conan/tools/cmake/toolchain/blocks.py
+++ b/conan/tools/cmake/toolchain/blocks.py
@@ -635,7 +635,7 @@ class ExtraFlagsBlock(Block):
 
 class CMakeFlagsInitBlock(Block):
     template = textwrap.dedent("""
-        foreach(config ${CMAKE_CONFIGURATION_TYPES})
+        foreach(config IN LISTS CMAKE_CONFIGURATION_TYPES)
             string(TOUPPER ${config} config)
             if(DEFINED CONAN_CXX_FLAGS_${config})
               string(APPEND CMAKE_CXX_FLAGS_${config}_INIT " ${CONAN_CXX_FLAGS_${config}}")

--- a/conans/test/functional/toolchains/cmake/test_cmake_toolchain.py
+++ b/conans/test/functional/toolchains/cmake/test_cmake_toolchain.py
@@ -271,8 +271,8 @@ def test_cmaketoolchain_no_warnings():
     client.run("create dep")
     client.run("install .")
     build_type = "-DCMAKE_BUILD_TYPE=Release" if platform.system() != "Windows" else ""
-    client.run_command("cmake . -DCMAKE_TOOLCHAIN_FILE=./conan_toolchain.cmake {}"
-                       "-Werror=dev --warn-uninitialized".format(build_type))
+    client.run_command("cmake -Werror=dev --warn-uninitialized . {}"
+                       " -DCMAKE_TOOLCHAIN_FILE=./conan_toolchain.cmake".format(build_type))
     assert "Using Conan toolchain" in client.out
     # The real test is that there are no errors, it returns successfully
 

